MapboxNavigation/NavigationMapView.swift

```diff
@@ -1335,3 +1346,95 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
         enableFrameByFrameCourseViewTracking(for: 3)
     }
 }
+
+// MARK: - Building Extrusion Highlighting
+
+extension NavigationMapView {
+       
+    /**
+     Receives coordinates for searching the map for buildings. If buildings are found, they will be highlighted in 2D or 3D depending on the `in3D` value.
+     
+     - parameter coordinates: Coordinates which represent buildings locations.
+     - parameter extrudesBuildings: Switch which allows to highlight buildings in either 2D or 3D. Defaults to true.
+     */
+    public func highlightBuildings(at coordinates: [CLLocationCoordinate2D], in3D extrudesBuildings: Bool = true) {
+        highlightBuildings(with: Set(coordinates.compactMap({ buildingIdentifier(at: $0) })), in3D: extrudesBuildings)
+    }
+    
+    /**
+     Removes the highlight from all buildings highlighted by `highlightBuildings(at:in3D:)`.
+     */
+    public func unhighlightBuildings() {
+        guard let highlightedBuildingsLayer = style?.layer(withIdentifier: StyleLayerIdentifier.buildingExtrusion) else { return }
+        
+        style?.removeLayer(highlightedBuildingsLayer)
+    }
+    
+    private func addBuildingsLayer() -> MGLFillExtrusionStyleLayer? {
+        if let highlightedBuildingsLayer = style?.layer(withIdentifier: StyleLayerIdentifier.buildingExtrusion) as? MGLFillExtrusionStyleLayer { return highlightedBuildingsLayer }
+        guard let buildingsSource = style?.source(withIdentifier: "composite") else { return nil }
+        
+        let highlightedBuildingsLayer = MGLFillExtrusionStyleLayer(identifier: StyleLayerIdentifier.buildingExtrusion, source: buildingsSource)
+        highlightedBuildingsLayer.sourceLayerIdentifier = "building"
+        highlightedBuildingsLayer.fillExtrusionColor = NSExpression(forConstantValue: buildingDefaultColor)
+        highlightedBuildingsLayer.fillExtrusionOpacity = NSExpression(forConstantValue: 0.05)
+        highlightedBuildingsLayer.fillExtrusionHeightTransition = MGLTransition(duration: 0.8, delay: 0)
+        highlightedBuildingsLayer.fillExtrusionOpacityTransition = MGLTransition(duration: 0.8, delay: 0)
+        
+        style?.addLayer(highlightedBuildingsLayer)
+        
+        return highlightedBuildingsLayer
+    }
+
+    private func buildingIdentifier(at coordinate: CLLocationCoordinate2D) -> Int64? {
+        let screenCoordinate = convert(coordinate, toPointTo: self)
+
+        // To increase a chance of selecting correct building identifier filter out
+        // features which do not contain appropriate attributes.
+        let features = visibleFeatures(at: screenCoordinate).filter {
+            let extrude = $0.attribute(forKey: "extrude") as? String
+            let type = $0.attribute(forKey: "type") as? String
+            return extrude == "true" && type == "building"
```

Comment:
@langsmith pointed out that `type` isn’t necessarily `building` for all buildings. In fact, the values are rather freeform, coming from OpenStreetMap’s [`building`](https://wiki.openstreetmap.org/wiki/Key:building) key.

This check is here because the call to `visibleFeatures(at:)` doesn’t filter the results to a specific style layer upfront, and the returned features don’t indicate which style layer they come from: mapbox/mapbox-gl-native#7162. The current road name label’s feature querying code solves this issue by figuring out upfront which style layers contain features from the Mapbox Streets source.

/ref https://github.com/mapbox/mapbox-navigation-ios/pull/2535#discussion_r468106341

Reply:
From https://github.com/mapbox/mapbox-navigation-ios/pull/2535#discussion_r468106341:

> Filter `MGLStyle.layers` to layers whose `sourceLayerIdentifier` is [`building`](https://docs.mapbox.com/vector-tiles/reference/mapbox-streets-v8/#building), then pass those layers into `visibleFeatures(at:styleLayerIdentifiers:)`. But this assumes the style shows buildings to begin with, which isn’t the case for Satellite Streets or some other likely styles.

Showing buildings is now a precondition of this method, so this approach is now feasible.

Reply:
Moved to first approach to mirror Android implementation, also removed check type = 'building' to increase chances of extrusion.